cpp/src/arrow/compute/exec/tpch_node.cc

```diff
@@ -0,0 +1,3836 @@
+                success &= GenerateTerminator(offset, rng, arr);
+                break;
+            case 4:
+                success &= GenerateNounPhrase(offset, rng, arr);
+                success &= GeneratePrepositionalPhrase(offset, rng, arr);
+                success &= GenerateVerbPhrase(offset, rng, arr);
+                success &= GenerateNounPhrase(offset, rng, arr);
+                success &= GenerateTerminator(offset, rng, arr);
+                break;
+            case 5:
+                success &= GenerateNounPhrase(offset, rng, arr);
+                success &= GeneratePrepositionalPhrase(offset, rng, arr);
+                success &= GenerateVerbPhrase(offset, rng, arr);
+                success &= GeneratePrepositionalPhrase(offset, rng, arr);
+                success &= GenerateTerminator(offset, rng, arr);
+                break;
+            default:
+                Unreachable("Random number should be between 0 and 5 inclusive");
+                break;
+            }
+            return success;
+        }
+
+        using GenerateColumnFn = std::function<Status(size_t)>;
+        class PartAndPartSupplierGenerator
+        {
+        public:
+            Status Init(
+                size_t num_threads,
+                int64_t batch_size,
+                float scale_factor)
+            {
+                if(!inited_)
+                {
+                    inited_ = true;
+                    batch_size_ = batch_size;
+                    scale_factor_ = scale_factor;
+
+                    thread_local_data_.resize(num_threads);
+                    for(ThreadLocalData &tld : thread_local_data_)
+                    {
+                        // 5 is the maximum number of different strings we need to concatenate
+                        tld.string_indices.resize(5 * batch_size_);
+                    }
+                    part_rows_to_generate_ = static_cast<int64_t>(scale_factor_ * 200000);
+                }
+                return Status::OK();
+            }
+            
+            int64_t part_batches_generated() const
+            {
+                return part_batches_generated_.load();
+            }
+
+            int64_t partsupp_batches_generated() const
+            {
+                return partsupp_batches_generated_.load();
+            }
+
+            Result<std::shared_ptr<Schema>> SetPartOutputColumns(const std::vector<std::string> &cols)
+            {
+                return SetOutputColumns(cols, part_types_, part_name_map_, part_cols_);
+            }
+
+            Result<std::shared_ptr<Schema>> SetPartSuppOutputColumns(const std::vector<std::string> &cols)
+            {
+                return SetOutputColumns(cols, partsupp_types_, partsupp_name_map_, partsupp_cols_);
+            }
+
+            Result<util::optional<ExecBatch>> NextPartBatch()
+            {
+                size_t thread_index = thread_indexer_();
```

Comment:
This is one of the great travesties of our time. Since the two ExecNodes have separate thread indexers, different threads may be mapped to the same index, which is a big problem. Ideally the thread index would be a thread_local variable, so that we don't have to grab a mutex to a hash table just to know which index our thread has. The technology just isn't there yet it seems.

Reply:
> This is one of the great travesties of our time.

Then we should fix it.  Once #12468 is merged we can assert that all work is done within the executor for the plan.  Then we can do something like: https://github.com/apache/arrow/commit/e2324141e57f8c2a3ed7b0da19e2d4103c64df7c

Would that work?

Reply:
Yes, as long as the thread indexes within an exec plan are consistent. We should maybe have a dedicated discussion for this elsewhere, I think Michal also had some thoughts on something like this.